core/message_handler_test.go

```diff
@@ -0,0 +1,230 @@
+	"math/big"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/0xProject/0x-mesh/constants"
+	"github.com/0xProject/0x-mesh/meshdb"
+	"github.com/0xProject/0x-mesh/zeroex"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMessageSharingIsolated(t *testing.T) {
+	meshDB, err := meshdb.New("/tmp/leveldb_testing/" + uuid.New().String())
+	defer meshDB.Close()
+	require.NoError(t, err)
+
+	testCases := []struct {
+		orderCount int
+		nextOffset int
+		max        int
+	}{
+		{
+			orderCount: 1,
+			nextOffset: 0,
+			max:        1,
+		},
+		{
+			orderCount: 2,
+			nextOffset: 0,
+			max:        1,
+		},
+		{
+			orderCount: 2,
+			nextOffset: 1,
+			max:        1,
+		},
+		{
+			orderCount: 2,
+			nextOffset: 1,
+			max:        2,
+		},
+		{
+			orderCount: 10,
+			nextOffset: 5,
+			max:        2,
+		},
+		{
+			orderCount: 10,
+			nextOffset: 1,
+			max:        2,
+		},
+		{
+			orderCount: 3,
+			nextOffset: 1,
+			max:        4,
+		},
+		{
+			orderCount: 3,
+			nextOffset: 20,
+			max:        4,
+		},
+	}
+
+	for _, testCase := range testCases {
+		orders := randomOrders(t, testCase.orderCount)
+
+		selector := &OrderSelector{
+			nextOffset: testCase.nextOffset,
+			db:         meshDB,
+		}
+
+		// Insert the orders into the database
+		insertOrders(t, selector, orders)
+
+		// Ensure that the correct orders are shared
+		verifyRoundRobinSharing(t, selector, testCase.nextOffset, testCase.max)
+
+		// Delete the orders from the database
+		deleteOrders(t, selector, orders)
+	}
+}
+
+func TestMessagesSharedSerial(t *testing.T) {
+	var allOrders []*meshdb.Order
+	meshDB, err := meshdb.New("/tmp/leveldb_testing/" + uuid.New().String())
+	defer meshDB.Close()
+	require.NoError(t, err)
+
+	selector := &OrderSelector{
+		nextOffset: 0,
+		db:         meshDB,
+	}
+
+	// Add five orders to the database
+	orders := randomOrders(t, 5)
+	allOrders = orders
+	insertOrders(t, selector, orders)
+	verifyRoundRobinSharing(t, selector, 0, 3)
+
+	// Add seven more orders to the database
+	orders = randomOrders(t, 7)
+	allOrders = append(allOrders, orders...)
+	insertOrders(t, selector, orders)
+	verifyRoundRobinSharing(t, selector, 3, 10)
+
+	// Delete three orders from the database
+	deleteOrders(t, selector, allOrders[0:3:3])
+	verifyRoundRobinSharing(t, selector, selector.nextOffset, 5)
+
+	// Add 12 more orders to the database
+	orders = randomOrders(t, 12)
+	allOrders = append(allOrders, orders...)
+	insertOrders(t, selector, orders)
+	verifyRoundRobinSharing(t, selector, selector.nextOffset, 7)
+}
+
+// Verify that the correct messages are shared by `GetMessagesToShare` given `orders`, a `nextOffset`, and `max`
+func verifyRoundRobinSharing(t *testing.T, selector *OrderSelector, nextOffset int, max int) {
+	notRemovedFilter := selector.db.Orders.IsRemovedIndex.ValueFilter([]byte{0})
+	ordersSnapshot, err := selector.db.Orders.GetSnapshot()
```

Comment:
I don't think we need a snapshot here. [The documentation](https://godoc.org/github.com/0xProject/0x-mesh/db#Snapshot) explains:

> Snapshot is a frozen, read-only snapshot of a DB state at a particular point in time.

It's only required when other goroutines might make modifications to the database and you want to run several queries without having the state change in between.